lapack/gonum/dlagv2.go

```diff
@@ -0,0 +1,209 @@
+func (impl Implementation) Dlagv2(a []float64, lda int, b []float64, ldb int, alphar, alphai, beta []float64) (csl, snl, csr, snr float64) {
+	const safmin = dlamchS
+	const ulp = dlamchP
+
+	// Shorthand float64 routines.
+	var (
+		max = math.Max
+		abs = math.Abs
+	)
+	// Scale A.
+	anorm := max(abs(a[0*lda+0])+abs(a[1*lda+0]),
+		abs(a[0*lda+1])+abs(a[1*lda+1]))
+	anorm = max(anorm, safmin)
+	ascale := 1. / anorm
+	a[0*lda+0] *= ascale
+	a[0*lda+1] *= ascale
+	a[1*lda+0] *= ascale
+	a[1*lda+1] *= ascale
+
+	// Scale B.
+	bnorm := max(abs(b[0*ldb+0]),
+		abs(b[0*ldb+1])+abs(b[1*ldb+1]))
+	bnorm = max(bnorm, safmin)
+	bscale := 1. / bnorm
+	b[0*lda+0] *= bscale
+	b[0*lda+1] *= bscale
+	b[1*lda+1] *= bscale
+
+	bi := blas64.Implementation()
+	// Values returned by Dlag2 used to calculate rotation matrix.
+	var wi, wr1, scale1 float64
+	switch {
+	// Check if A can be deflated.
+	case abs(a[1*lda+0]) <= ulp:
+		// wi,snr,snl = 0,0,0 (are already zero)
+		csl, csr = 1, 1
+		a[1*lda+0] = 0
+		b[1*lda+0] = 0
+
+		// Check if B is singular.
+	case abs(b[0*lda+0]) <= ulp:
+		// snr,wi = 0,0 (are already zero)
+		csl, snl, _ = impl.Dlartg(a[0*lda+0], a[1*lda+0])
+		csr = 1
+		bi.Drot(2, a[0*lda+0:], 1, a[1*lda+0:], 1, csl, snl)
+		bi.Drot(2, b[0*ldb+0:], 1, b[1*ldb+0:], 1, csl, snl)
+		a[1*lda+0] = 0
+		b[0*ldb+0] = 0
+		b[1*ldb+0] = 0
+
+	case abs(b[1*ldb+1]) <= ulp:
+		// snl,wi = 0,0 (are already zero)
+		csr, snr, _ = impl.Dlartg(a[1*lda+1], a[1*lda+0])
+		snr *= -1
+		bi.Drot(2, a[0*lda+0:], lda, a[0*lda+1:], lda, csr, snr)
+		bi.Drot(2, b[0*ldb+0:], ldb, b[0*ldb+1:], ldb, csr, snr)
+		csl = 1
+		a[1*lda+0] = 0
+		b[1*ldb+0] = 0
+		b[1*ldb+1] = 0
+
+		// B is nonsingular, first compute the eigenvalues of (A,B).
+	default:
+		// scale1, _, wr1, _, wi = impl.Dlag2(a, lda, b, ldb)
+		csl, snl, csr, snr = impl.dlagv2aux(a, lda, b, ldb, scale1, wr1, wi)
+
+	}
+
+	// Unscaling.
+	a[0*lda+0] *= anorm
+	a[0*lda+1] *= anorm
+	a[1*lda+0] *= anorm
+	a[1*lda+1] *= anorm
+
+	b[0*ldb+0] *= bnorm
+	b[0*ldb+1] *= bnorm
+	b[1*ldb+0] *= bnorm
+	b[1*ldb+1] *= bnorm
+
+	if wi == 0 {
+		alphar[0] = a[0*lda+0]
+		alphar[1] = a[1*lda+1]
+		alphai[0] = 0
+		alphai[1] = 0
+		beta[0] = b[0*ldb+0]
+		beta[1] = b[1*ldb+1]
+	} else {
+		alphar[0] = anorm * wr1 / scale1 / bnorm
+		alphai[0] = anorm * wi / scale1 / bnorm
+		alphar[1] = alphar[0]
+		alphai[1] = -alphai[0]
+		beta[0] = 1
+		beta[1] = 1
+	}
+
+	return csl, snl, csr, snr
+}
+
+// dlagv2aux solves for eigenvalues of (A,B) using first eigenvalue
+// returned by Dlag2.
+//
+// dlagv2aux is an internal auxiliary routine used in exported Implementation.Dlagv2.
+func (impl Implementation) dlagv2aux(a []float64, lda int, b []float64, ldb int, scale1, wr1, wi float64) (csl, snl, csr, snr float64) {
+	bi := blas64.Implementation()
+
+	// Shorthand float64 routines.
+	var (
+		max = math.Max
+		abs = math.Abs
+	)
+
+	// A pair of complex conjugate eigenvalues: first compute the SVD of the matrix B.
+	if wi != 0 {
```

Comment:
Why change the order?

Reply:
Instead of having a `if else` for real and imaginary eigenvalues there's an `if` for the "special" case of imaginary eigenvalues.

This dedents all the real eigenvalue case making it slightly more readable since there are a couple of `if` in there as well. It also helps that `dlagv2` is it's own method and not contained within the `switch default`. The reference is heavily indented because of the nested `if`s.